data_model/src/query/mod.rs

```diff
@@ -31,199 +23,207 @@ use self::{
-        FindAccountKeyValueByIdAndKey(FindAccountKeyValueByIdAndKey),
-        FindAccountsByDomainId(FindAccountsByDomainId),
-        FindAccountsWithAsset(FindAccountsWithAsset),
-        FindAllAssets(FindAllAssets),
-        FindAllAssetsDefinitions(FindAllAssetsDefinitions),
-        FindAssetById(FindAssetById),
-        FindAssetDefinitionById(FindAssetDefinitionById),
-        FindAssetsByName(FindAssetsByName),
-        FindAssetsByAccountId(FindAssetsByAccountId),
-        FindAssetsByAssetDefinitionId(FindAssetsByAssetDefinitionId),
-        FindAssetsByDomainId(FindAssetsByDomainId),
-        FindAssetsByDomainIdAndAssetDefinitionId(FindAssetsByDomainIdAndAssetDefinitionId),
+    pub enum IterableQueryBox {
+        FindAllDomains(IterableQueryWithFilterFor<FindAllDomains>),
+        FindAllAccounts(IterableQueryWithFilterFor<FindAllAccounts>),
+        FindAllAssets(IterableQueryWithFilterFor<FindAllAssets>),
+        FindAllAssetsDefinitions(IterableQueryWithFilterFor<FindAllAssetsDefinitions>),
+        FindAllRoles(IterableQueryWithFilterFor<FindAllRoles>),
+
+        FindAllRoleIds(IterableQueryWithFilterFor<FindAllRoleIds>),
+        FindPermissionsByAccountId(IterableQueryWithFilterFor<FindPermissionsByAccountId>),
+        FindRolesByAccountId(IterableQueryWithFilterFor<FindRolesByAccountId>),
+        FindTransactionsByAccountId(IterableQueryWithFilterFor<FindTransactionsByAccountId>),
+        FindAccountsWithAsset(IterableQueryWithFilterFor<FindAccountsWithAsset>),
+
+        FindAllPeers(IterableQueryWithFilterFor<FindAllPeers>),
+        FindAllActiveTriggerIds(IterableQueryWithFilterFor<FindAllActiveTriggerIds>),
+        FindAllTransactions(IterableQueryWithFilterFor<FindAllTransactions>),
+        FindAllBlocks(IterableQueryWithFilterFor<FindAllBlocks>),
+        FindAllBlockHeaders(IterableQueryWithFilterFor<FindAllBlockHeaders>),
+    }
+
+    /// An enum of all possible iterable query batches.
+    ///
+    /// We have an enum of batches instead of individual elements, because it makes it easier to check that the batches have elements of the same type and reduces serialization overhead.
+    #[derive(
+        Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema, FromVariant,
+    )]
+    pub enum IterableQueryOutputBatchBox {
+        Domain(Vec<Domain>),
+        Account(Vec<Account>),
+        Asset(Vec<Asset>),
+        AssetDefinition(Vec<AssetDefinition>),
+        Role(Vec<Role>),
+        Parameter(Vec<Parameter>),
+        Permission(Vec<Permission>),
+        Transaction(Vec<TransactionQueryOutput>),
+        Peer(Vec<Peer>),
+        RoleId(Vec<RoleId>),
+        TriggerId(Vec<TriggerId>),
+        Block(Vec<SignedBlock>),
+        BlockHeader(Vec<BlockHeader>),
+    }
+
+    /// An enum of all possible singular queries
+    #[derive(
+        Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema, FromVariant,
+    )]
+    pub enum SingularQueryBox {
```

Comment:
from what I can see, it seems to me that, in the end, only `FindExecutorDataModel` will remain as a singular query? Is that indeed the case?

Reply:
Yes, I think with projections implemented and some refactors to the queries, all the other ones will be representable as iterable queries. Although this assumes "smart" projections, like "get metadata value with key `42`" and not just field accesses.

Reply:
which means that there will no longer be any dichotomy between singular and iterable queries. It doesn't make sense to have 1 singular query.

This is the reason why I want to subjugate defaults to iterable queries already now, so we don't change api later